command/var_put.go

```diff
@@ -0,0 +1,472 @@
+	case len(args) == 0:
+		// no-op
+	case isArgStdinRef(args[0]):
+		verbose(fmt.Sprintf("Creating variable %q using specification from stdin", path))
+		stat, _ := os.Stdin.Stat()
+		if (stat.Mode() & os.ModeCharDevice) == 0 {
+			c.contents, err = io.ReadAll(os.Stdin)
+			if err != nil {
+				c.Ui.Error(fmt.Sprintf("Error reading from stdin: %s", err))
+				return 1
+			}
+		}
+		args = args[1:]
+
+	case isArgFileRef(args[0]):
+		arg := args[0]
+		verbose(fmt.Sprintf("Creating variable %q from specification file %q", path, arg))
+		fPath := arg[1:]
+		c.contents, err = os.ReadFile(fPath)
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf("error reading %q: %s", fPath, err))
+			return 1
+		}
+		args = args[1:]
+	default:
+		// no-op - should be KV arg
+	}
+
+	sv, err := c.makeVariable(path)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Failed to parse variable data: %s", err))
+		return 1
+	}
+
+	if len(args) > 0 {
+		data, err := parseArgsData(stdin, args)
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf("Failed to parse K=V data: %s", err))
+			return 1
+		}
+
+		for k, v := range data {
+			vs := v.(string)
+			if vs == "" {
+				if _, ok := sv.Items[k]; ok {
+					c.Ui.Warn(fmt.Sprintf("Removed item %q", k))
+					delete(sv.Items, k)
+				} else {
+					c.Ui.Warn(fmt.Sprintf("Item %q does not exist, continuing...", k))
+				}
+				continue
+			}
+			sv.Items[k] = vs
+		}
+	}
+	// Get the HTTP client
+	client, err := c.Meta.Client()
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error initializing client: %s", err))
+		return 1
+	}
+
+	if enforce {
+		sv.ModifyIndex = checkIndex
+	}
+
+	if forceOverwrite {
+		sv, _, err = client.Variables().Update(sv, nil)
+	} else {
+		sv, _, err = client.Variables().CheckedUpdate(sv, nil)
+	}
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error creating variable: %s", err))
+		return 1
+	}
+
+	verbose(fmt.Sprintf("Created variable %q with modify index %v", sv.Path, sv.ModifyIndex))
+
+	var out string
+	switch c.outFmt {
+	case "json":
+		out = sv.AsJSON()
+	case "pretty-json":
+		out = sv.AsPrettyJSON()
+	case "hcl":
+		out = renderAsHCL(sv)
+	case "go-template":
+		if out, err = renderWithGoTemplate(sv, c.tmpl); err != nil {
+			c.Ui.Error(err.Error())
+			return 1
+		}
+	case "table":
+		// the renderSVAsUiTable func writes directly to the ui and doesn't error.
+		renderSVAsUiTable(sv, c)
+		return 0
+	default:
+		return 0
+	}
+	c.Ui.Output(out)
+	return 0
+}
+
+// makeVariable creates a variable based on whether or not there is data in
+// content and the format is set.
+func (c *VarPutCommand) makeVariable(path string) (*api.Variable, error) {
+	var err error
+	out := new(api.Variable)
+	if len(c.contents) == 0 {
+		out.Path = path
+		out.Namespace = c.Meta.namespace
+		out.Items = make(map[string]string)
+		return out, nil
+	}
+	switch c.inFmt {
+	case "json":
+		err = json.Unmarshal(c.contents, out)
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshaling json: %w", err)
```

Comment:
We chatted previously about why automatically detecting the input format is challenging. But it might be nice in the case where there's a parsing error if we could suggest trying a different input format? Here's what passing HCL to the default format looks like:

```
$ nomad var put -namespace prod yet/another/path @spec.nsv.hcl
Failed to parse variable data: error unmarshaling json: invalid character '#' looking for beginning of value
```